--[[

TODO: Create modules for each of the following
- Signal (to replace sleitnick's signal)
- Heap (to replace Trove)
- Spring (maybe update epoxyish?)
- EffectPlayer (to play effects from vfx suite)
- BadgeUtils
- Bezier
- Ragdoll
- Binder (binds class to roblox instance)
- TracebackUtils (utilities for debug.traceback)
- MarketplaceUtils (for purchases)
- ProgressionUtils (for currency/score tracking, quests, boosters, spending, etc)

Also include the following third-party modules:
- Promise
- Cmdr

--]]

local Praxis = {}

return Praxis